pkg/cmd/server/start/start_node.go

```diff
@@ -281,13 +320,295 @@ func (o NodeOptions) CreateNodeConfig() error {
+	if err != nil {
+		return err
+	}
+	privateKeyData := utilcert.EncodePrivateKeyPEM(privateKey)
+	ipAddresses, dnsNames := crypto.IPAddressesDNSNames(hostnames)
+	csrData, err := utilcert.MakeCSR(privateKey, &pkix.Name{
+		CommonName: dnsNames[0],
+	}, dnsNames, ipAddresses)
+	if err != nil {
 		return err
 	}
 
+	signingRequest := &certificates.CertificateSigningRequest{
+		ObjectMeta: kapi.ObjectMeta{
+			GenerateName: fmt.Sprintf("node-bootstrapper-%s-server-", safeSecretName(o.NodeArgs.NodeName)),
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: csrData,
+		},
+	}
+
+	csr, err := c.Certificates().CertificateSigningRequests().Create(signingRequest)
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(serverKeyPath, privateKeyData, 0600); err != nil {
+		return err
+	}
+
+	err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		if deny := hasCSRCondition(csr.Status.Conditions, certificates.CertificateDenied); deny != nil {
+			glog.V(2).Infof("Bootstrap signing rejected (%s): %s", deny.Reason, deny.Message)
+			return false, fmt.Errorf("certificate signing request rejected (%s): %s", deny.Reason, deny.Message)
+		}
+		if approved := hasCSRCondition(csr.Status.Conditions, certificates.CertificateApproved); approved != nil {
+			glog.V(2).Infof("Bootstrap serving cert approved")
+			return true, nil
+		}
+		csr, err = c.Certificates().CertificateSigningRequests().Get(csr.Name)
+		return false, err
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(serverCertPath, csr.Status.Certificate, 0600); err != nil {
+		return err
+	}
+	if _, err := tls.LoadX509KeyPair(serverCertPath, serverKeyPath); err != nil {
+		return fmt.Errorf("bootstrap server certificate does not match private key: %v", err)
+	}
+
 	return nil
 }
 
+func (o NodeOptions) loadBootstrap(hostnames []string, nodeConfigDir string) error {
+	if err := os.MkdirAll(nodeConfigDir, 0700); err != nil {
+		return err
+	}
+
+	kubeClientConfig, err := o.NodeArgs.KubeConnectionArgs.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+	var c kclientset.Interface
+	c, err = kclientset.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Bootstrapping from API server %s (experimental)", kubeClientConfig.Host)
+
+	c, err = o.loadBootstrapClientCertificate(nodeConfigDir, c, 1*time.Minute)
+	if err != nil {
+		return err
+	}
+	if err := o.loadBootstrapServerCertificate(nodeConfigDir, hostnames, c, 1*time.Minute); err != nil {
+		return err
+	}
+
+	nodeClientCAPath := filepath.Join(nodeConfigDir, "node-client-ca.crt")
+	if len(kubeClientConfig.CAData) > 0 {
+		if err := ioutil.WriteFile(nodeClientCAPath, []byte(kubeClientConfig.CAData), 0600); err != nil {
+			return err
+		}
+	}
+
+	// try to refresh the latest node-config.yaml
+	o.ConfigFile = filepath.Join(nodeConfigDir, "node-config.yaml")
+	config, err := c.Core().ConfigMaps("openshift-infra").Get("node-config")
```

Comment:
Figure out whether we need complex layering experimentally, have a good default, come back over time.  Would like some practical experience with this and ansible.